### PR TITLE
Factor out mutation-setting logic from 'Connection'.

### PR DESCRIPTION
**Pull request by tseaver**
Toward #514.

### CODE REVIEW

File: gcloud/datastore/batch.py [hunk 1]
```diff
@@ -209,15 +210,7 @@ def put(self, entity):
...
-        key_pb = entity.key.to_protobuf()
-        properties = dict(entity)
-        exclude = tuple(entity.exclude_from_indexes)
-
-        self.connection.save_entity(
-            self.dataset_id, key_pb, properties,
-            exclude_from_indexes=exclude, mutation=self.mutation)
-
-        if entity.key.is_partial:
+        if _assign_entity_to_mutation(self.mutation, entity):
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: Small nit: the name `_assign_entity_to_mutation` doesn't seem to indicate the return value will mean auto-id assigned. Would it be more intuitive to pass `self._auto_id_entities` as an argument?

File: gcloud/datastore/batch.py [hunk 2]
```diff
@@ -279,3 +271,35 @@ def __exit__(self, exc_type, exc_val, exc_tb):
                 self.rollback()
         finally:
             _BATCHES.pop()
+
+
+def _assign_entity_to_mutation(mutation_pb, entity):
+    """Helper method for ``Batch.put``."""
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: Needs documentation for inputs and return value.

File: gcloud/datastore/batch.py [hunk 3]
```diff
@@ -279,3 +271,35 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+    auto_id = entity.key.is_partial
+
+    key_pb = entity.key.to_protobuf()
+    key_pb = helpers._prepare_key_for_request(key_pb)
+
+    if auto_id:
+        insert = mutation_pb.insert_auto_id.add()
+    else:
+        insert = mutation_pb.upsert.add()
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: Can you add a comment above here explaining that we don't use `insert` or `update` but are aware of their existence and may add in future?

File: gcloud/datastore/test_batch.py [hunk 4]
```diff
@@ -305,17 +347,6 @@ class _Connection(object):
     def __init__(self, *new_keys):
         self._commit_result = _CommitResult(*new_keys)
         self._committed = []
-        self._saved = []
-        self._deleted = []
-
-    def save_entity(self, dataset_id, key_pb, properties,
```
Comment (dhermes) [inline gcloud/datastore/test_batch.py]: Yay!

**Comment by dhermes** [comment]
@tseaver 

1. This is just a part of the third bullet in #514? I.E., slowly reduce dependencies on `save_entity()` and `delete_entities()`?
1. ISTM `delete()` and `put()` are ready to go in `datastore.api` as you outlined in #518, yes? I can do that if you like or just sit back and watch.

Mostly LGTM other than small comments

**Comment by tseaver** [comment]
@dhermes: rebased / updated to address your concerns in 8bd5059.  PTAL.

**Comment by dhermes** [comment]
One [concern][1] still dangling at

```python
     if auto_id:
         insert = mutation_pb.insert_auto_id.add()
         auto_id_entities.append(entity)
     else:
         insert = mutation_pb.upsert.add()
```

> Can you add a comment above here explaining that we don't use `insert` or `update`
> but are aware of their existence and may add in future?

---

Can you address my two questions [above][2]? See #527 regarding point 2) above.

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/pull/524#discussion_r22691464
[2]: https://github.com/GoogleCloudPlatform/gcloud-python/pull/524#issuecomment-69269359

**Comment by tseaver** [comment]
Good catch:  0525530 has the comment, which I hadn't saved to disk :(.

WRT #527: I really don't want to have `begin()` moved outside the context manager:  it removes most of the point of the CM in the first place.

**Comment by dhermes** [comment]
LGTM

**Pull request merged**